### PR TITLE
update grpc trace interceptor

### PR DESCRIPTION
**Pull request by singhravidutt**

### CODE REVIEW

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientGrpcTracingInterceptor.java [hunk 1]
```diff
@@ -16,60 +16,87 @@
...
+  }
+
+  @NonNull
+  static String fmtProto(@NonNull WriteObjectRequest msg) {
+    if (msg.hasChecksummedData()) {
+      ByteString content = msg.getChecksummedData().getContent();
+      WriteObjectRequest.Builder b = msg.toBuilder();
+      ByteString snip = ByteString.copyFromUtf8(String.format("<size (%d)>", content.size()));
+      b.getChecksummedDataBuilder().setContent(snip);
+      return b.build().toString();
```
[the remark this answers is not in the document]
Comment (singhravidutt) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientGrpcTracingInterceptor.java:76]: pretty big, for read/write it contains data of whole chunk. That's the reason we are overriding it with size of content.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientGrpcTracingInterceptor.java [hunk 2]
```diff
@@ -16,60 +16,87 @@
...
-            method,
-            callOptions.withStreamTracerFactory(
-                new ClientStreamTracer.Factory() {
-                  @Override
-                  public ClientStreamTracer newClientStreamTracer(
-                      StreamInfo info, Metadata headers) {
-                    return streamTracer;
-                  }
-                }))) {
+    TrackingStreamTracer streamTracer = new TrackingStreamTracer(rpcMethodName);
```
[the remark this answers is not in the document]
Comment (singhravidutt) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientGrpcTracingInterceptor.java:98]: Shouldn't add much. I don't have benchmarking numbers on that Give it is enabled only on request I wouldn't worry much. Although would have a tpch run and to see how much extra time it takes.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientGrpcTracingInterceptor.java [hunk 3]
```diff
@@ -106,268 +134,95 @@ public void onClose(Status status, Metadata trailers) {
...
     public void traceRequestMessage(MessageLite message) {
-      logRequestMessage(message);
-      updateRequestCounter();
+      logger.atInfo().log(
+          "%s",
+          toJson(
+              getRequestTrackingInfo()
+                  .put(
+                      GoogleCloudStorageTracingFields.REQUEST_MESSAGE_AS_STRING.name,
+                      fmtProto(message))
```
[the remark this answers is not in the document]
Comment (singhravidutt) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientGrpcTracingInterceptor.java:156]: It is corelatable based on invocationIds. A grpc stream may have many messages exchanged and not all of them have all the information i.e. All WriteObject request msg don't carry information about the object against which upload is being done but the very first msg do have that information. But all the messages in a stream are stitched together using the invocationId.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientGrpcTracingInterceptor.java [hunk 4]
```diff
@@ -16,60 +16,87 @@
...
+      b.getChecksummedDataBuilder().setContent(snip);
+      return b.build().toString();
+    }
+    return msg.toString();
+  }
+
+  @NonNull
+  static String fmtProto(@NonNull ReadObjectResponse msg) {
+    if (msg.hasChecksummedData()) {
+      ByteString content = msg.getChecksummedData().getContent();
```
[the remark this answers is not in the document]
Comment (singhravidutt) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientGrpcTracingInterceptor.java:84]: Given `ReadObjectResponse` and `WriteObjectRequest` don't share a common interface with `hasChecksummedData` it becomes really difficult to merge these function. I do see one optimization where updated content can be fetched from same function. Will add that. Open to suggections.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientGrpcTracingInterceptor.java [hunk 5]
```diff
@@ -16,60 +16,87 @@
...
-  private static final String DEFAULT_INVOCATION_ID = "NOT-FOUND";
+  private static final DateTimeFormatter dtf =
+      DateTimeFormatter.ofPattern("yyyy/MM/dd HH:mm:ss.SSS");
   private static final Metadata.Key<String> idempotencyKey =
       Metadata.Key.of(IDEMPOTENCY_TOKEN_HEADER, Metadata.ASCII_STRING_MARSHALLER);
 
+  @NonNull
+  static String fmtProto(@NonNull Object obj) {
+    if (obj instanceof WriteObjectRequest) {
+      return fmtProto((WriteObjectRequest) obj);
```
[the remark this answers is not in the document]
Comment (singhravidutt) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientGrpcTracingInterceptor.java:54]: Do have it covered as part of integration test case`GoogleCloudStorageClientInterceptorIntegrationTest.testWriteLogs`, not sure why it's not showing up in codecov.

**Comment by singhravidutt** [comment]
test passed for https://pantheon.corp.google.com/cloud-build/builds;region=global/7c24aa32-d205-49d2-8b99-fd01d5d615f3;step=3?e=13803378&mods=dataproc_env_prod&project=cloud-dataproc-ci

**Review by arunkumarchacko — APPROVED** [review]

**Pull request merged**